Relationship columns definition

pxpm opened the pull request:
Relationship column definitions:

### BelongsTo
```php
        $this->crud->addColumn([
            'name' => 'category.name',
            'key' => 'my_own_category_second',
            'type' => 'text',
            'label' => 'forced text'
        ]);
        $this->crud->addColumn([
            'name' => 'category.created_at',
            'type' => 'datetime'
        ]);
        $this->crud->addColumn('category');
        $this->crud->addColumn('category.name');
        $this->crud->addColumn('category.parent.name');
        $this->crud->addColumn([
            'name' => 'category',
            'key' => 'my_own_category',
            'attribute' => 'name',
            'type' => 'select',
            'label' => 'select arr def'
        ]);
```
![belongsTo](https://user-images.githubusercontent.com/7188159/110802283-f3eb4a00-8275-11eb-9c6f-d4ef93491434.png)

### BelongsToMany

```php
$this->crud->addColumn('categories.name');
$this->crud->addColumn('categories');
$this->crud->addColumn('categories.parent.name');
```
![belongsToMany](https://user-images.githubusercontent.com/7188159/110802370-08c7dd80-8276-11eb-93eb-ae70f9791303.png)

### HasOne

```php
$this->crud->addColumn('address');
$this->crud->addColumn('address.country');
$this->crud->addColumn('address.icon');
$this->crud->addColumn('address.icon.name');
```
![hasOne](https://user-images.githubusercontent.com/7188159/110802536-2eed7d80-8276-11eb-9a93-e2a9d9d9b67d.png)

### HasMany

```php
$this->crud->addColumn('postalboxes');
$this->crud->addColumn('postalboxes.postal_name');
$this->crud->addColumn('postalboxes.monster.text');
```
![hasMany](https://user-images.githubusercontent.com/7188159/110802681-504e6980-8276-11eb-840d-0a656e559d6f.png)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/73a65192-1c09-4fef-9191-910ca94f3f9a) was created.

tabacitu commented:
@pxpm  I just LOVE how you've detailed how to use each relationship type here.

Is this a breaking change though?

pxpm commented:
@tabacitu I DO NOT think this is a BC. All `old` methods are supported, plus the ones provided here. Also I don't think this breaks `searchLogic`.

The only `open possibility` for a BC is when instead of getting the column `by name`, we get the column by `generated key`. At that time it's not yet generated,  `CRUD::column('some_column')`, developer provide only column name, but later the `key` will be generated, so we can generate it in advance and check if any column with that key already exists. 

I'v detailed here all the tests I'v done to add relationship columns, do you think people might be using more ways of adding it ? 

Let me know,
Pedro

promatik reviewed (approved):
Great PR @pxpm 👏
I tested, works great. I can't see how this can be a BC.

tabacitu commented on src/app/Library/CrudPanel/CrudColumn.php:
```diff
@@ -32,7 +32,9 @@ class CrudColumn
     public function __construct($name)
     {
-        $column = $this->crud()->firstColumnWhere('name', $name);
+        $generated_key = str_replace('.', '__', $name);
+
+        $column = $this->crud()->firstColumnWhere('key', $generated_key);
```
```suggestion
        $column = $this->crud()->firstColumnWhere('key', str_replace('.', '__', $name));
```

tabacitu commented on tests/Unit/CrudPanel/CrudPanelColumnsTest.php:
```diff
@@ -127,16 +127,17 @@ class CrudPanelColumnsTest extends BaseDBCrudPanelTest
         ],
         'accountDetails__nickname' => [
-            'name'        => 'accountDetails.nickname',
+            'name'        => 'accountDetails',
             'label'       => 'AccountDetails.nickname',
             'type'        => 'relationship',
             'key'         => 'accountDetails__nickname',
             'priority'    => 1,
+            'attribute' => 'nickname',
             'tableColumn' => false,
             'orderable'   => false,
             'searchLogic' => false,
             'relation_type' => 'HasOne',
-            'entity' => 'accountDetails.nickname',
+            'entity' => 'accountDetails',
             'model' => 'Backpack\CRUD\Tests\Unit\Models\AccountDetails',
         ],
```
Hmmm what about the old test, with `accountDetails.nickname`? Does that stop working?

pxpm replied:
Indeed. That's what we are fixing here. 

That test was added when I added the code in `ColumnsProtectedMethods.php` that I am refactoring now. 

I added that way because of the `key` that was generated later so I kept the `attribute` in the `name` and `entity`. 

That lead to other unexpected consequences thus I now refactored `key` to be generated first and remove the actual attribute from the relation string, as it technically is not part of the relation string anyway, it's just what we want to show.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/27e12de2-7008-4194-9aeb-cd07727d5320) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/d807687e-4c1f-43d6-85f4-87a0a9325c24) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/27e12de2-7008-4194-9aeb-cd07727d5320) completed: **1 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/d807687e-4c1f-43d6-85f4-87a0a9325c24) completed: **1 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/eac05802-7147-4167-91f9-a8ea77c08e53) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/c5fa632a-5d80-4aab-bb50-aa8a2ad06904) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/c5fa632a-5d80-4aab-bb50-aa8a2ad06904) completed: **1 updated code elements**

tabacitu commented:
I think we should consider this a breaking change, and merge it in 4.2. Since the `key` is an actual feature that Columns provide, people might be using it. And in this PR we're changing how that works. That'll definitely affect very few people, but I think this might break some admin panels and that's a very bad UX.

4.2.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/1817f844-b90f-4e16-95ff-bd00d82e747f) was created.

pxpm commented:
Hello @tabacitu !

I've updated this PR to be a non-BC. 

I've removed the changes in the tests, and with key generation, I think it's a non-breaking now. 

Can you give it a shoot ? 

Thanks,
Pedro

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/67c362e7-646e-4dd4-95c0-3ba543f6e09b) completed: **2 updated code elements**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/0650eb26-3003-4ec0-9761-d9255912120b) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/22f604e0-1806-479c-8919-772c70472988) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/59a33835-cc84-4dd6-88e5-f7edca85fe6a) was created.

tabacitu commented:
Thanks for the nudge @axelzuzek 🙏   Merged into `master`, it'll be available after we tag a version 4.1.59 (most likely on Monday). Until then, you can use this branch (by requiring `dev-relationship-columns-definition`) or `dev-master`.

Thanks again guys!

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/59a33835-cc84-4dd6-88e5-f7edca85fe6a) completed: **2 new issues, 2 updated code elements**

tabacitu commented:
Done - tagged. A `composer update backpack/crud` should fix it for you @axelzuzek

realtebo commented:
Thanks for this. Just a question: docs have been updated?

tabacitu commented:
Not yet - I'll leave that up to @pxpm 😀🙏
